### ISSUE TITLE
Process the asynchronous Kademlia results and verify that the artifact manage puts torrents in the DHT

### ISSUE THREAD

**Issue by mgrand** [pyrsia/pyrsia#306]
This is the conclusion of the work described in #281. The issue is to add a mechanism for asynchronously receiving event notifications when a kademlia operation finished. Once that is done we will finish the artifact manager unit test that now just checks that putting the torrent in the DHT was initiated without error. It will also verify that the torrent has actually been added to the DHT.

**Comment by mgrand** [comment]
To facilitate delivery of Kademlia events by one thread to another thread that is matched to the event by a query ID, I am creating the `message_delivery` module.

**Comment by mgrand** [comment]
To be able to do unit testing with Kademlia live, I need to change the way the libp2p swarm is initialized. Currently it is initialized from main. It will need to be statically initialized instead.

**Comment by mgrand** [comment]
There was some legacy code that was using channels to receive Kademlia messages. I had to change the way all that worked. 

Now that I have tired unit testing, I have discovered that Kademlia does not work unless the polling loop is running. Right now the polling loop is running from main, which does not work for unit tests. I will move the polling look to the SwarmSingleThreadedProxy, adding a start_polling_loop method to that.

**Comment by mgrand** [comment]
@betarelease During the planning meeting you suggested that I focus on untangling first. Thank you for the suggestion. I will spilt the event loop into one for libp2p events and another for everything else. That way, I don't have to take on the rest of the concerns right now. My "unit" test will exercise Kademlia and just Kademlia.

**Comment by mgrand** [comment]
Working on getting tests to pass, I have discovered that the logging create we are using, env_logging, buffers its output. I discovered this because I have a bug that is causing a test to hang. I am no seeing any of its log output unless I comment out the async call that causes the hang.

I will be investigating using other logging crates, at least for unit tests. When I have identified a suitable alternative, I will post in the team slack channel asking for opinions.
